log4j-api/src/main/java/org/apache/logging/log4j/util/ProviderUtil.java

```diff
@@ -163,6 +172,88 @@ static void lazyInit() {
         }
     }
 
+    /**
+     * Used to test the public {@link #getProvider()} method.
+     */
+    static Provider selectProvider(final PropertiesUtil properties) {
```

Comment:
I presume you sufficiently covered this method with tests. :see_no_evil:

Reply:
I tested the externally observable behavior:

1. the `log4j.provider` system property creates a provider with the specified class name,
2. the `log4j2.loggerContextFactory` system property create a provider with the specified logger context factory class,
3. if more than one implementation is available, the implementation with highest `getPriority` wins.

What happens in the remaining cases (multiple system properties set, typos in the names of properties) are in my opinion implementation details and we can change them at will without breaking the `#getProvider()` contract.

Reply:
I increased the test coverage with some smoke tests.

If users specify a faulty configuration, I just check that a warning or error appears and no exception is thrown.